### PR TITLE
Detection of 'native' CPU

### PR DESCRIPTION
**Pull request by fxcoudert**
Here's a first attempt calling `sysctlbyname("hw.cpufamily", …)`
At some point, we might want to fill in the blanks for older models, but definitely no hurry :)

### CODE REVIEW

**Comment by fxcoudert** [comment]
Some added notes on https://github.com/iains/gcc-darwin-arm64/commit/43a9cd85686ff5f7446a1129bc1ebf54826e9f59 (aarch64, Darwin: Initial implementation of Apple CPU cores).

- For Apple cores, we don't care about the `IMP` and `PART` arguments of the `AARCH64_CORE` macro. They're only used for core identification from /proc/cpuinfo, which we are bypassing.
- Alternatively, we could reuse these fields to carry the values of `hw.cpufamily` (M1, M2, etc) and `hw.cpusubfamily` (M1, M1 Pro, etc). But I think a simple switch is easier to read and maintain at this stage.

**Comment by iains** [comment]
I was looking at the big.LITTLE stuff (I.e. where we have performance and efficiency cores), but I do not think that there's any way that we can action it at present.

As far as we might make action at present, the difference between M1 and M2 could be useful (M2 has support for BTI and BF16) but the BTI case is a bit complex since the aarch64 backend assumes it comes with armv8.5 and M2 is _mostly_ armv8.5 but it is missing one mandatory feature (according to the Arm ref. manual) so I've made it as armv8.4+additions.

**Comment by iains** [comment]
do you think this is something we should try to get into our gcc-14 release?
(if so, it presumably needs rebasing)

**Comment by fxcoudert** [comment]
OK. We should probably add the M3 there. But it's a rat's race, and we should have a sensible default. What happens if someone passes `native` and we don't detect the sysctl value?

**Comment by iains** [comment]
> OK. We should probably add the M3 there. 

I'll add M3 to the known CPUs shortly (done locally, but I need to rebase and push)

> But it's a rat's race, and we should have a sensible default. What happens if someone passes `native` and we don't detect the sysctl value?

that's a general question about error recovery/fallback - what does the linux impl. do if it encounters an unknown cpu in /proc/.... ?

we could have a fallback to m1, I suppose ..  getting early access to h/w info seems unlikely given that it's likely to be held close prior to any release.

**Comment by fxcoudert** [comment]
- Rebased, 
- Added M3 variants (all named M3 because I don't think we want to have more detail)
- Removed older, unused variants

Can't test right now, sorry

**Comment by iains** [comment]
response from IRC:  hmm it should only return null on error, when we simply didn't find something we know it returns DEFAULT_ARCH + extensions https://sourcegraph.com/github.com/gcc-mirror/gcc@7d8585c0c0e5934780281abdee256ae6553e56e8/-/blob/gcc/config/aarch64/driver-aarch64.cc?L402

I will take a look at this when adding the patch to the branch (I know you are also busy).

File: gcc/config/aarch64/driver-aarch64.cc [hunk 1]
```diff
@@ -28,6 +28,95 @@
...
+  /* Are we processing -march, mtune or mcpu?  */
+  arch = strcmp (argv[0], "arch") == 0;
+  if (!arch)
+    tune = strcmp (argv[0], "tune") == 0;
+  if (!arch && !tune)
+    cpu = strcmp (argv[0], "cpu") == 0;
+  if (!arch && !tune && !cpu)
+    return NULL;
+
+  sysctlbyname("hw.cpufamily", &family, &len, NULL, 0);
```
Comment (iains) [inline gcc/config/aarch64/driver-aarch64.cc:68]: I think we should check that the sysctlbyname succeeds and issue a diagnostic if not.  For the sake of "doing the same thing as Linux" returning NULL would also be reasonable there.

**Review by iains — COMMENTED** [review]
I've squashed the three patches here onto my current rebase (which should be ready tomorrow - without making any other changes).

**Comment by iains** [comment]
https://github.com/iains/gcc-darwin-arm64/commit/9bc66f9b11baf1c291b2afac429cb12bf8461f91 includes the content of this branch.

**Comment by iains** [comment]
the merged patch has been tested on M1 and M3 - so I think we're all good with this - and maybe a couple of minor tweaks to the code for error checking.

**Comment by fxcoudert** [comment]
The other uses of this `sysctlbyname` in the darwin code of GCC expect that it never fails, so I think we're safe, but I'll open a PR to add error checking at some point.

**Comment by iains** [comment]
I have merged the patch here and added one to select `apple-m1` as a default CPU if there is no configure one.

**Pull request closed without merging**